### ISSUE TITLE
Investigate use cases for nested query builders

### ISSUE THREAD

**Issue by jakeboone02** [react-querybuilder/react-querybuilder#515]
~~Example using v6.3.0: https://codesandbox.io/s/snowy-snow-ub85bl?file=/src/CustomRule.tsx~~

Updated with v8.1.0: https://codesandbox.io/p/devbox/q6jr4c?file=%2Fsrc%2FCustomRule.tsx

Based on [this Discord chat](https://discord.com/channels/1020373981645250660/1020373982127591526/1106256798870151259).

![image](https://github.com/react-querybuilder/react-querybuilder/assets/366438/781f967a-42d3-4684-a7d5-810f0fda561f)

### Use cases

- SQL `EXISTS` clauses

### Misc. thoughts

- Store the query from the nested query builder in the `value` property of the parent rule.
- Nested query builders should inherit all props from their parent query builder by default?
- Store additional/override props for the nested query builder in the `nested` property of fields or operators.
- Trigger the display of a nested query builder based on truthiness of `nested` property, or maybe a new value source selection (`"query"`?).

### Questions

- What should `formatQuery` do with rules that have nested queries? Can they be integrated somehow like a sub-group, or would they always need a custom `ruleProcessor`?
  - Use case for `EXISTS` clauses would need some more thought. Initially:
    ```sql
    WHERE EXISTS (SELECT 1 FROM [???] WHERE [nested query rules here])
    --                   ^?       ^?         ^ the easy part, maybe
    ```

**Comment by jakeboone02** [comment]
### Ideas

- Store the query from the nested query builder in the `value` property of the parent rule.
- Nested query builder should inherit all props from parent query builder by default?
- Store additional/override props for the nested query builder in the `nested` property of fields or operators.
- Trigger the display of a nested query builder based on truthiness of `nested` property, or maybe a new value source selection (`"query"`?).

### Questions

- What should `formatQuery` do with rules that have nested queries? Can they be integrated somehow like a sub-group, or would they always need a custom `ruleProcessor`?

**Comment by jakeboone02** [comment]
## Where things stand after #900

#900 says "Addresses" rather than "Fixes" because it covers the _self-contained nested collection_ half of this issue, not the rest.

**Delivered by #900:**

- `match: { mode, threshold }` on rules; the subquery lives in the rule's `value` as a `RuleGroupType` (the "store the query in `value`" idea from the OP).
- `matchModes` / `subproperties` field props, `getMatchModes` / `getSubQueryBuilderProps` QB props, `MatchModeEditor` control (incl. native).
- Modes: `all`, `some`, `none`, `atLeast`, `atMost`, `exactly`.
- Export: full support in jsonlogic, jsonata, cel, spel, natural_language, mongodb_query; partial in sql / parameterized / drizzle (postgres + primitive arrays only) and elasticsearch (`some`/`none` only).
- Import: `parseCEL` and `parseJsonLogic` only.

**Still open:**

1. **Correlated subqueries over a _separate_ source** — the headline use case here: `WHERE EXISTS (SELECT 1 FROM otherTable WHERE otherTable.fk = main.id AND <subquery rules>)`. #900 only unnests a column of the current row (`unnest(field)`, `$elemMatch`, jsonlogic `some`). There's no way to express the `FROM [???]` source or the correlation predicate — the two `^?` marks in the OP remain unanswered.
2. **Trigger mechanism** — the OP floated a `nested` field/operator prop or a `"query"` value source. #900 requires selecting a match mode, and there's no per-_operator_ subquery config.
3. **Per-field subquery builder props** — only the QB-level `getSubQueryBuilderProps` callback exists today; the OP wanted overrides on the field/operator definition.
4. **Export coverage** — nothing for `parameterized_named`, `prisma`, `sequelize`, `tanstack_db`, `ldap`, legacy `mongodb`; partial for sql/parameterized/drizzle (postgres-only, primitive arrays) and elasticsearch (no `all`/threshold modes).
5. **Parser coverage** — `parseSQL` ignores `EXISTS`/`unnest` (the grammar already emits `SubQuery`/`hasExists` nodes, currently unused), `parseMongoDB` ignores `$elemMatch`, `parseJSONata`/`parseSpEL` ignore their array-predicate forms.

## Proposed approach

Layer a **subquery source** concept on top of the existing `match` machinery rather than building a parallel feature. `match` stays the **sole** trigger for subquery semantics. A rule targeting a foreign source keeps `match` and `value` (a `RuleGroupType`) and gains source/correlation metadata resolved from the _field definition_ — not stored in the query object, so serialized queries stay portable and small.

New field props:

```ts
type CorrelationPair = { parentField: string; subField: string };

type Correlation =
  | CorrelationPair    // the 95% case
  | CorrelationPair[]  // implicit AND, composite keys
  | RuleGroupType;     // advanced escape hatch
```

plus `subquerySource` (table/collection name). `subproperties` continues to describe the fields available inside the subquery, and field/operator-level `subQueryBuilderProps` get merged under the existing `getSubQueryBuilderProps` result.

The shorthand forms are the documented API; `RuleGroupType` stays supported only because it's nearly free (rule processors already recurse into a `RuleGroupType` in `rule.value`, and the group processors already serialize groups), not because arbitrary join predicates are expected to be common.

Rule processors branch three ways: no `match` → today's scalar path; `match` with no resolvable source → today's array/unnest path; `match` + resolvable source → correlated subquery. Correlated SQL is standard-compliant, so that third path drops the postgres-only restriction and unlocks `parameterized_named`, `prisma`/`sequelize`, and `tanstack_db` where feasible.

Because the metadata is deliberately _not_ persisted in the query, `formatQuery` can't emit correlated output from the query object alone — but that needs no new API, since `FormatQueryOptions.fields` already exists. Worth documenting explicitly; when the source can't be resolved, the rule is dropped silently, consistent with existing unsupported-rule behavior.

## The part that actually needs designing: aliasing and column qualification

This is the real blocker, and it isn't a types question. Today `defaultRuleProcessorSQL` does:

```ts
// TODO?: Randomize this alias
const arrayElementAlias = 'elem_alias';

const nestedArrayFilter = defaultRuleGroupProcessorSQL(
  transformQuery(rule.value as RuleGroupType, {
    ruleProcessor: r => ({ ...r, field: arrayElementAlias }),
  }),
  opts
);
```

Every rule inside the subquery has its `field` **replaced wholesale** by one hardcoded alias. That's correct for `unnest()` over an array of primitives — there's exactly one column — but it means a subquery rule can't reference a distinct column at all today, let alone a column of the outer row.

Correlated subqueries need:

- distinct parent and child aliases, stable across nesting depth, instead of one constant;
- **qualification** of subquery rules (`child.col`) rather than replacement, preserving each rule's own field name and honoring `quoteFieldNamesWith` / `fieldIdentifierSeparator`;
- both-side qualification of the correlation predicate (`child.fk = parent.id`) — there's currently no mechanism for anything inside a subquery to refer outward;
- an alias stack, so subquery-within-subquery keeps the outer alias reachable.

Leaning toward deterministic aliases by depth (`t0`, `t1`, …): depth-uniqueness resolves the existing `TODO?: Randomize` collision concern without the snapshot/`dbquery` churn that randomization would cause.

## Two ideas from the OP I'm now inclined to drop

**A `'query'` value source.** I'd initially argued that `ValueSource` growing to `'value' | 'field' | 'expression' | 'parameter'` made this well-precedented, since `'expression'` already stores a structured, core-uninterpreted payload in `value`. On closer look that precedent doesn't transfer: `'expression'` resolves to a **scalar-producing** operand, and its support lives in the opt-in `@react-querybuilder/expr` package behind a processor registry — core processors never branch on it. A `'query'` source would have to be handled in core, in every processor, and would shift `valueSource` from meaning "where does the value come from" to "what shape is this rule." `match` already answers the latter, so a second marker buys no expressive power, creates round-trip asymmetry with parsers, and — since back-compat requires `match`-only rules to work forever anyway — adds a trigger path instead of replacing one.

One consequence to verify if this stands: the scalar-vs-subquery toggle then lives entirely in `MatchModeEditor`, so a field with `matchModes` must still be able to represent "no match mode selected" and produce an ordinary scalar rule.

**A UI source selector.** The source should be determined by the field (`Customers → Orders`, `Invoice → Payments`). Letting the user pick an arbitrary table starts turning the component into a SQL designer.

Similarly, the `getSubquerySource` / `getCorrelation` callbacks I floated are deferred — nothing in the initial use cases needs dynamic resolution, and they can be added later mirroring `getMatchModesUtil` if demand appears.

## Roadmap

Representation + every exporter + every parser + ORM integrations in one change would be an enormous PR. Splitting it:

**Phase 1 — representation + SQL proof.** Prove `query → internal model → SQL` end to end, then stop. Alias/qualification model; `subquerySource` + `Correlation` types; resolution utils alongside `getMatchModesUtil`; `sql` + `parameterized` + `parameterized_named` processors; hook plumbing; `defaultValidator` treating an empty correlated subquery group as invalid; unit + `dbquery` tests at 100% coverage; docs (`tips/subqueries.mdx`, support matrix, prop tables, CHANGELOG).

**Phase 2 — formatter fan-out.** drizzle (`exists()`), mongodb_query, elasticsearch (`all`/threshold via `nested` + `must_not` inversion or script counts), prisma/sequelize/tanstack_db where a native construct exists, plus a decide-and-document pass on `ldap` and legacy `mongodb`.

**Phase 3 — parsers.** `parseSQL` `EXISTS`, `parseMongoDB` `$elemMatch`, `parseJSONata`, `parseSpEL`. The SQL grammar already emitting `hasExists` is encouraging, but this is more than visitor plumbing — the visitor has to separate the correlation predicate from the subquery's own filter predicates using table aliases. Import has historically been harder than export here, so no parity promises.

## Notes

- Backward compatibility is mandatory: every new field/QB prop is optional, and absent them output is byte-identical to today. Targets a minor release.
- Nested-within-nested subqueries should work by construction — tests, but no special-casing. The phase 1 alias stack is what makes that true.
- Formats that can't express correlated subqueries should keep dropping such rules silently rather than throwing.
- Interaction with the `expression`/`parameter` value sources needs definition: whether correlation predicates may be `ExpressionNode`s, and whether rules inside a subquery group may use `lhs` expressions (default: yes, by construction).
- **Longer term, explicitly out of scope.** A rule was originally `{ field, operator, value }`; post-#900 it's really `{ field, operator, match?, value }` where `value` polymorphically carries scalars, `ExpressionNode`s, or `RuleGroupType`s. If this area keeps growing (joins, `EXISTS`, quantified comparisons), an explicit operand model — `{ field, operator, rhs: ScalarOperand | QueryOperand }` — would likely be cleaner than continuing to overload `value`. Massively breaking, so not proposed now, but worth recording that the original "every rule has a scalar value" assumption is being stretched. It's also another reason not to paper over the gap with a `'query'` value source.

## Open questions

1. Correlation shorthand naming: `{ parentField, subField }`, `{ parentField, childField }`, or something aligned with the existing `subproperties` terminology?
2. Does the existing array/unnest path migrate to the new deterministic alias scheme, or keep `elem_alias` to avoid churn in existing snapshots?
3. Is `subquerySource` alone enough to make a field subquery-capable, or must `matchModes` also be set? (Related: what does a correlated rule with no match mode mean — plain `EXISTS`, or invalid?)
4. For `prisma`/`sequelize`, are relation filters (`some`/`every`/`none`) in scope, or documented-unsupported? Semantically close but not identical; the model shouldn't be contorted to fit Prisma.
5. Threshold modes on correlated subqueries: emit `(select count(*) ...) >= n` for all dialects, or restrict to dialects covered by the `dbquery` suites?
6. Is the `'query'` value source permanently rejected, or parked pending a future explicit operand model?
